### PR TITLE
Mule 11912

### PR DESCRIPTION
**Pull request by alegmarra**

### CODE REVIEW

File: core/src/main/java/org/mule/runtime/core/api/NestedProcessor.java [hunk 1]
```diff
@@ -1,49 +0,0 @@
...
-package org.mule.runtime.core.api;
-
-import java.util.Map;
-
-/**
- * Callback interface used by message processors methods.
- * <p/>
- * The method parameters of type {@link NestedProcessor} will be able to receive other message processors.
- */
-public interface NestedProcessor {
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/api/NestedProcessor.java]: ❤️

File: core/src/main/java/org/mule/runtime/core/el/mvel/MessageVariableResolverFactory.java [hunk 2]
```diff
@@ -30,7 +30,7 @@
   public static final String PAYLOAD = "payload";
   public static final String ATTRIBUTES = "attributes";
   public static final String MESSAGE_PAYLOAD = MESSAGE + "." + PAYLOAD;
-  public static final String FLOW_VARS = "flowVars";
+  public static final String FLOW_VARS = "vars";
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/el/mvel/MessageVariableResolverFactory.java]: this is probably what broke MEL. This needs to be rolled back

File: core/src/main/java/org/mule/runtime/core/internal/processor/EventBasedResult.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+import org.mule.runtime.api.message.Message;
+import org.mule.runtime.api.metadata.MediaType;
+import org.mule.runtime.api.metadata.TypedValue;
+import org.mule.runtime.core.api.InternalEvent;
+import org.mule.runtime.extension.api.runtime.operation.Result;
+
+import java.util.Optional;
+
+public class EventBasedResult<T, A> extends Result<T, A> {
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/internal/processor/EventBasedResult.java]: EventedResult?

Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/internal/processor/EventBasedResult.java]: final

File: core/src/main/java/org/mule/runtime/core/internal/processor/EventBasedResult.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  }
+
+  @Override
+  public T getOutput() {
+    return (T) event.getMessage().getPayload().getValue();
+  }
+
+  @Override
+  public Optional<A> getAttributes() {
+    return Optional.ofNullable((A) event.getMessage().getAttributes().getValue());
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/internal/processor/EventBasedResult.java]: static import

File: core/src/main/java/org/mule/runtime/core/internal/processor/chain/ImmutableProcessorChainExecutor.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+/**
+ * An implementation of {@link Chain} that wraps a {@link Processor} and allows to execute it
+ *
+ * @since 4.0
+ */
+public class ImmutableProcessorChainExecutor implements Chain, Initialisable {
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/internal/processor/chain/ImmutableProcessorChainExecutor.java]: why is this here instead of extension-support? If the answer is "to keep core for exporting things that shouldn't" is a valid answer. Just making sure the answer is in fact valid

File: core/src/main/java/org/mule/runtime/core/internal/processor/chain/ImmutableProcessorChainExecutor.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        originalEvent.getContext().error(e);
+      }
+    }
+
+    return null;
+  }
+
+  @Override
+  public void initialise() throws InitialisationException {
+    LifecycleUtils.initialiseIfNeeded(chain);
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/internal/processor/chain/ImmutableProcessorChainExecutor.java]: static import

File: core/src/main/java/org/mule/runtime/core/internal/processor/chain/ImmutableProcessorChainExecutor.java [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+  private Consumer<Result> successHandler;
+  private BiConsumer<Throwable, Result> errorHandler;
+
+  /**
+   * Creates a new immutable instance
+   *
+   * @param event the original {@link InternalEvent} for the execution of the given chain
+   * @param chain a {@link Processor} chain to be executed
+   */
+  public ImmutableProcessorChainExecutor(InternalEvent event,
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/internal/processor/chain/ImmutableProcessorChainExecutor.java]: @dfeist will appreciate your input on this one

File: core/src/main/java/org/mule/runtime/core/internal/processor/simple/SetPayloadMessageProcessor.java [hunk 8]
```diff
@@ -22,6 +23,7 @@
  */
 public class SetPayloadMessageProcessor extends SimpleMessageProcessor {
 
+  private static final TypedValue NULL_TYPED_VALUE = new TypedValue<>(null, OBJECT);
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/internal/processor/simple/SetPayloadMessageProcessor.java]: boy scout rule :)

File: modules/extensions-spring-support/Heisenberg_result.xsd [hunk 9]
```diff
@@ -0,0 +1,2362 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/Heisenberg_result.xsd]: what is this?

File: modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/capability/xml/schema/builder/SourceSchemaDelegate.java [hunk 10]
```diff
@@ -49,10 +48,16 @@ private void registerSourceElement(SourceModel sourceModel, String typeName, Dsl
     element.setName(dslSyntax.getElementName());
     element.setType(new QName(builder.getSchema().getTargetNamespace(), typeName));
     element.setAnnotation(builder.createDocAnnotation(sourceModel.getDescription()));
-    element.setSubstitutionGroup(MULE_ABSTRACT_MESSAGE_SOURCE);
+    element.setSubstitutionGroup(getSourceSubstitutionGroup(sourceModel));
     builder.getSchema().getSimpleTypeOrComplexTypeOrGroup().add(element);
   }
 
+  private QName getSourceSubstitutionGroup(SourceModel sourceModel) {
+    return sourceModel.getStereotype().equals(MuleStereotypeFactory.source())
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/capability/xml/schema/builder/SourceSchemaDelegate.java]: in which case would a source have a different stereotype?

File: modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/config/dsl/DefaultExtensionBuildingDefinitionProvider.java [hunk 11]
```diff
@@ -159,6 +160,11 @@ public void onConfiguration(ConfigurationModel model) {
           }
 
           @Override
+          protected void onConstruct(ConstructModel model) {
+            //TODO
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/config/dsl/DefaultExtensionBuildingDefinitionProvider.java]: do

File: modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/config/dsl/ExtensionDefinitionParser.java [hunk 12]
```diff
@@ -211,6 +214,29 @@ protected ExtensionDefinitionParser(Builder baseDefinitionBuilder, DslSyntaxReso
    */
   protected abstract Builder doParse(Builder definitionBuilder) throws ConfigurationException;
 
+  protected void parseNestedComponents(List<? extends NestableElementModel> nestedComponents) {
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/config/dsl/ExtensionDefinitionParser.java]: seems like this whole method is a left over

File: modules/extensions-spring-support/src/test/java/org/mule/runtime/module/extension/internal/capability/xml/DefaultExtensionSchemaGeneratorTestCase.java [hunk 13]
```diff
@@ -172,6 +171,14 @@
         .collect(toList());
   }
 
+  private boolean shouldUpdateExpectedFilesOnError() {
+    // Utility to batch fix input files when severe model changes are introduced.
+    // Use carefully, not a mechanism to get away with anything.
+    // First check why the generated json is different and make sure you're not introducing any bugs.
+    // This should NEVER be committed as true
+    return false;
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/test/java/org/mule/runtime/module/extension/internal/capability/xml/DefaultExtensionSchemaGeneratorTestCase.java]: excess of chetness

File: modules/extensions-spring-support/src/test/java/org/mule/runtime/module/extension/internal/capability/xml/DefaultExtensionSchemaGeneratorTestCase.java [hunk 14]
```diff
@@ -172,6 +171,14 @@
         .collect(toList());
   }
 
+  private boolean shouldUpdateExpectedFilesOnError() {
+    // Utility to batch fix input files when severe model changes are introduced.
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/test/java/org/mule/runtime/module/extension/internal/capability/xml/DefaultExtensionSchemaGeneratorTestCase.java]: this should be the method's jdoc then

File: modules/extensions-spring-support/src/test/java/org/mule/runtime/module/extension/internal/resources/ExtensionModelJsonGeneratorTestCase.java [hunk 15]
```diff
@@ -143,6 +144,14 @@
         .collect(toList());
   }
 
+  private boolean shouldUpdateExpectedFilesOnError() {
+    // Utility to batch fix input files when severe model changes are introduced.
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/test/java/org/mule/runtime/module/extension/internal/resources/ExtensionModelJsonGeneratorTestCase.java]: as jdoc

File: modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/ScopeExecutionTestCase.java [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+
+  @Test
+  public void exceptionOnCallbacksFailure() throws Exception {
+    expectedException.expect(IllegalArgumentException.class);
+    expectedException.expectMessage("ON_ERROR_EXCEPTION");
+    runFlow("exceptionOnCallbacksFailure");
+  }
+
+  @Test
+  @org.junit.Ignore("TODO FOREACH")
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/ScopeExecutionTestCase.java]: replace message with jira issue

File: modules/extensions-spring-support/src/test/resources/typed-value-on-config-config.xml [hunk 17]
```diff
@@ -6,17 +6,17 @@
...
     </typed-value:config>
 
     <typed-value:config name="staticConfig" stringTypedValue="string">
         <typed-value:differed-door address="address"/>
     </typed-value:config>
 
     <flow name="typedValueOnDynamicConfig">
-        <set-variable variableName="address" value="address"/>
-        <set-payload value="string" mimeType="application/json" encoding="UTF-8"/>
+        <set-variable variableName="address" value="&quot;address&quot;" mimeType="application/json" encoding="UTF-8"/>
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/test/resources/typed-value-on-config-config.xml]: why these quotes? Sounds like a cover up for something that broke

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/enricher/StereotypesDeclarationEnricher.java [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+  public void enrich(ExtensionLoadingContext extensionLoadingContext) {
+    new EnricherDelegate().apply(extensionLoadingContext);
+  }
+
+  private static class EnricherDelegate {
+
+    private final Map<StereotypeDefinition, StereotypeModel> stereotypes = new HashMap<>();
+
+    public void apply(ExtensionLoadingContext extensionLoadingContext) {
+      ExtensionDeclarer extensionDeclarer = extensionLoadingContext.getExtensionDeclarer();
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/enricher/StereotypesDeclarationEnricher.java]: this method must be completely executed using the extension's class loader

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ParametersResolver.java [hunk 19]
```diff
@@ -104,15 +106,31 @@ public static ParametersResolver fromDefaultValues(ParameterizedModel parameteri
   public ResolverSet getParametersAsResolverSet(ParameterizedModel model, MuleContext muleContext) throws ConfigurationException {
 
     List<ParameterGroupModel> inlineGroups = getInlineGroups(model.getParameterGroupModels());
+    List<ParameterModel> nestedParameter = new ArrayList<>();
+    //    if(model instanceof ConstructModel){
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ParametersResolver.java]: delte

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ProcessorChainValueResolver.java [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+   * Returns a {@link Chain} that wraps the given {@link Processor} using the current {@code event}
+   *
+   * @param context a {@link ValueResolvingContext}
+   * @return a {@link Chain}
+   * @throws MuleException
+   */
+  @Override
+  public Chain resolve(ValueResolvingContext context) throws MuleException {
+    try {
+      muleContext.getRegistry().registerObject(new ObjectNameHelper(muleContext).getUniqueName(""), chain);
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ProcessorChainValueResolver.java]: why is this necessary? we can't register this each time the chain is resolved. 

We need to try to avoid this, and if there's no way to do that, then at least this should be idempotent and register the chain only once.

@dfeist can you advice on this being actually needed?

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ProcessorChainValueResolver.java [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+      initialiseIfNeeded(chain, muleContext);
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ProcessorChainValueResolver.java]: why do you need to initialise? If this goes into the registry, the registry should be the one doing the initialisation

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ResolverSet.java [hunk 22]
```diff
@@ -126,6 +126,13 @@ private Object resolveValue(ValueResolver<?> resolver, ValueResolvingContext con
     return value;
   }
 
+  public ResolverSet merge(ResolverSet resolverSet) {
+    ResolverSet newResolverSet = new ResolverSet(muleContext);
+    resolvers.forEach(newResolverSet::add);
+    resolverSet.getResolvers().forEach(newResolverSet::add);
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ResolverSet.java]: can't you addAll()?

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/DefaultExtensionModelFactoryTestCase.java [hunk 23]
```diff
@@ -87,17 +88,20 @@ public void flyweight() {
 
   @Test
   public void blockingExecutionTypes() {
-    ExtensionModel extensionModel = createExtension(HeisenbergExtension.class);
+    final List<String> blockingOperations = Arrays.asList("killMany", "executeAnything", "alwaysFailsWrapper",
```
Comment (marianogonzalez) [inline modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/DefaultExtensionModelFactoryTestCase.java]: why are scopes marked as blocking? Can't we have NB chains?

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/resolver/NestedProcessorValueResolverTestCase.java [hunk 24]
```diff
@@ -24,40 +25,54 @@
...
+    ProcessorChainValueResolver resolver = new ProcessorChainValueResolver(messageProcessor);
+    muleContext.getInjector().inject(resolver);
+    final InternalEvent event = testEvent();
 
-    ArgumentCaptor<InternalEvent> captor = ArgumentCaptor.forClass(InternalEvent.class);
-    verify(messageProcessor).process(captor.capture());
+    Chain nestedProcessor = resolver.resolve(ValueResolvingContext.from(event));
+    nestedProcessor
+        .onSuccess(result -> {
+          assertThat(result.getOutput(), is(TEST_PAYLOAD));
```
Comment (marianogonzalez) [inline modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/resolver/NestedProcessorValueResolverTestCase.java]: is this actually testing anything? I don't see you calling the process() method. Plus, in which thread will these assertions be executed?

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/resolver/NestedProcessorValueResolverTestCase.java [hunk 25]
```diff
@@ -24,40 +25,54 @@
...
 
   @Test
   public void alwaysGivesDifferentInstances() throws Exception {
-    NestedProcessorValueResolver resolver = new NestedProcessorValueResolver(messageProcessor);
-    resolver.setMuleContext(muleContext);
-    NestedProcessor resolved1 = resolver.resolve(ValueResolvingContext.from(testEvent()));
-    NestedProcessor resolved2 = resolver.resolve(ValueResolvingContext.from(testEvent()));
+    ProcessorChainValueResolver resolver = new ProcessorChainValueResolver(messageProcessor);
+    muleContext.getInjector().inject(resolver);
+    Chain resolved1 = resolver.resolve(ValueResolvingContext.from(testEvent()));
```
Comment (marianogonzalez) [inline modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/resolver/NestedProcessorValueResolverTestCase.java]: This is very wrong. It should always be the same chain. What happens if I put a stateful operation in the scope? This is not compatible with an idempotent filter for example or with <ee:cache>

File: tests/test-extensions/heisenberg-extension/src/main/java/org/mule/test/heisenberg/extension/HeisenbergScopes.java [hunk 26]
```diff
@@ -0,0 +1,70 @@
...
+import java.util.Map;
+
+public class HeisenbergScopes {
+
+  @Throws(HeisenbergErrorTyperProvider.class)
+  public void killMany(@AllowedStereotypes({KillingStereotype.class, DrugKillingStereotype.class}) Chain killOperations,
+                       CompletionCallback<String, Void> callback, String reason)
+      throws Exception {
+    // TODO
+    callback.success(Result.<String, Void>builder().output(reason).build());
```
Comment (marianogonzalez) [inline tests/test-extensions/heisenberg-extension/src/main/java/org/mule/test/heisenberg/extension/HeisenbergScopes.java]: chain is not been executed.

File: tests/test-extensions/heisenberg-extension/src/main/java/org/mule/test/heisenberg/extension/HeisenbergScopes.java [hunk 27]
```diff
@@ -0,0 +1,70 @@
...
+import org.mule.test.heisenberg.extension.stereotypes.DrugKillingStereotype;
+import org.mule.test.heisenberg.extension.stereotypes.KillingStereotype;
+
+import java.util.Map;
+
+public class HeisenbergScopes {
+
+  @Throws(HeisenbergErrorTyperProvider.class)
+  public void killMany(@AllowedStereotypes({KillingStereotype.class, DrugKillingStereotype.class}) Chain killOperations,
+                       CompletionCallback<String, Void> callback, String reason)
```
Comment (marianogonzalez) [inline tests/test-extensions/heisenberg-extension/src/main/java/org/mule/test/heisenberg/extension/HeisenbergScopes.java]: not sure if I remember the final decision on this. Do I always need the completionCallback when using chains? Isn't the onSuccess() thing enough?

If I do need the CompletionCAllback, then:

a) That should be clear on the Chain jdoc
b) There should be a validator which fails if I get a chain without a callback
c) all scopes should appear as non blocking in the extension model

File: tests/test-extensions/heisenberg-extension/src/main/java/org/mule/test/heisenberg/extension/HeisenbergScopes.java [hunk 28]
```diff
@@ -0,0 +1,70 @@
...
+      throws Exception {
+    // TODO
```
Comment (marianogonzalez) [inline tests/test-extensions/heisenberg-extension/src/main/java/org/mule/test/heisenberg/extension/HeisenbergScopes.java]: todo what?

File: tests/test-extensions/heisenberg-extension/src/main/java/org/mule/test/heisenberg/extension/HeisenbergScopes.java [hunk 29]
```diff
@@ -0,0 +1,70 @@
...
+    callback.success(Result.<String, Void>builder().output(reason).build());
+  }
+
+  public void executeAnything(Chain chain, CompletionCallback<Void, Void> cb) {
+    chain.onSuccess(cb::success).process();
+  }
+
+  public void payloadModifier(Chain chain, CompletionCallback<Void, Void> cb,
+                              Object payload, Map<String, String> attributes) {
+    chain.onSuccess(cb::success).process(payload, attributes);
```
Comment (marianogonzalez) [inline tests/test-extensions/heisenberg-extension/src/main/java/org/mule/test/heisenberg/extension/HeisenbergScopes.java]: I would like another test in which the payload is actually mutated inside the onSucess()

File: core/src/main/java/org/mule/runtime/core/internal/processor/chain/ImmutableProcessorChainExecutor.java (hunk 5 above)
Comment (alegmarra) [reply core/src/main/java/org/mule/runtime/core/internal/processor/chain/ImmutableProcessorChainExecutor.java]: uses `InternalEvent` and `MessageProcessorChain`, but they are both exported already. Can be moved

File: modules/extensions-spring-support/Heisenberg_result.xsd (hunk 9 above)
Comment (alegmarra) [reply modules/extensions-spring-support/Heisenberg_result.xsd]: this, kids, is why we don't do `git add .`

File: modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/capability/xml/schema/builder/SourceSchemaDelegate.java (hunk 10 above)
Comment (alegmarra) [reply modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/capability/xml/schema/builder/SourceSchemaDelegate.java]: could be a child stereotype that inherits from Source but is more restricted. Maybe something like `PollingSource` and `ListenerSource`

File: modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/config/dsl/ExtensionDefinitionParser.java (hunk 12 above)
Comment (alegmarra) [reply modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/config/dsl/ExtensionDefinitionParser.java]: nope, used in OperationDefinitionParser

File: modules/extensions-spring-support/src/test/resources/typed-value-on-config-config.xml (hunk 17 above)
Comment (alegmarra) [reply modules/extensions-spring-support/src/test/resources/typed-value-on-config-config.xml]: nope, is the way to pass an string with mimeType json. Talked about this with stiv, is how typed values work

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/enricher/StereotypesDeclarationEnricher.java (hunk 18 above)
Comment (alegmarra) [reply modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/enricher/StereotypesDeclarationEnricher.java]: done

File: core/src/main/java/org/mule/runtime/core/internal/processor/chain/ImmutableProcessorChainExecutor.java (hunk 5 above)
Comment (marianogonzalez) [reply core/src/main/java/org/mule/runtime/core/internal/processor/chain/ImmutableProcessorChainExecutor.java]: if this can be internal then move it to internal. The future of InternalEvent and MessageProcessorChain is uncertain

File: modules/extensions-spring-support/src/test/resources/typed-value-on-config-config.xml (hunk 17 above)
Comment (marianogonzalez) [reply modules/extensions-spring-support/src/test/resources/typed-value-on-config-config.xml]: why did you need to change it then?

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/resolver/NestedProcessorValueResolverTestCase.java (hunk 25 above)
Comment (marianogonzalez) [reply modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/resolver/NestedProcessorValueResolverTestCase.java]: to me more clear, the MEssageprocessorChain must always be the same. But the SDK chain must always be a different instance

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ResolverSet.java (hunk 22 above)
Comment (alegmarra) [reply modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ResolverSet.java]: is `add(name, value)`

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/DefaultExtensionModelFactoryTestCase.java (hunk 23 above)
Comment (alegmarra) [reply modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/DefaultExtensionModelFactoryTestCase.java]: bad var name, these are nonBlockingOperations, check the assertion:
`assertThat(operation.isBlocking(), is(!nonBlockingOperations.contains(operationName)));`

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ResolverSet.java (hunk 22 above)
Comment (marianogonzalez) [reply modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ResolverSet.java]: putAll() then

File: tests/test-extensions/heisenberg-extension/src/main/java/org/mule/test/heisenberg/extension/HeisenbergScopes.java (hunk 29 above)
Comment (alegmarra) [reply tests/test-extensions/heisenberg-extension/src/main/java/org/mule/test/heisenberg/extension/HeisenbergScopes.java]: check `org.mule.test.heisenberg.extension.HeisenbergScopes#neverFailsWrapper`

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ProcessorChainValueResolver.java [hunk 30]
```diff
@@ -0,0 +1,84 @@
...
+
+  @Inject
+  private MuleContext muleContext;
+
+  public ProcessorChainValueResolver(MessageProcessorChain chain) {
+    this.chain = chain;
+  }
+
+  public ProcessorChainValueResolver(List<Processor> algo) {
+    chain = new DefaultMessageProcessorChainBuilder().chain(algo).build();
```
Comment (dfeist) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ProcessorChainValueResolver.java]: You can use `MesssageProcessors.newChain()`

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ProcessorChainValueResolver.java (hunk 20 above)
Comment (dfeist) [reply modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ProcessorChainValueResolver.java]: Was this resolved now?

File: core/src/main/java/org/mule/runtime/core/internal/processor/chain/ImmutableProcessorChainExecutor.java (hunk 7 above)
Comment (dfeist) [reply core/src/main/java/org/mule/runtime/core/internal/processor/chain/ImmutableProcessorChainExecutor.java]: @marianogonzalez anything specific? Concept/design of ImmutableProcessorChainExecutor or use of reactor? Nothing major jumps out at me..

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/resolver/NestedProcessorValueResolverTestCase.java (hunk 25 above)
Comment (alegmarra) [reply modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/resolver/NestedProcessorValueResolverTestCase.java]: see `mule/modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/ScopeExecutionTestCase#verifySameProcessorInstance`
